// @openzeppelin/upgrades-core/dist/eip-1967.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toEip1967Hash = exports.toFallbackEip1967Hash = exports.getBeaconAddress = exports.getImplementationAddress = exports.getAdminAddress = void 0;
const ethereumjs_util_1 = require("ethereumjs-util");
const bn_js_1 = __importDefault(require("bn.js"));
const provider_1 = require("./provider");
async function getAdminAddress(provider, address) {
    const storage = await getStorageFallback(provider, address, toEip1967Hash('eip1967.proxy.admin'), toFallbackEip1967Hash('org.zeppelinos.proxy.admin'));
    return parseAddress(storage);
}
exports.getAdminAddress = getAdminAddress;
async function getImplementationAddress(provider, address) {
    const storage = await getStorageFallback(provider, address, toEip1967Hash('eip1967.proxy.implementation'), toFallbackEip1967Hash('org.zeppelinos.proxy.implementation'));
    if (isEmptySlot(storage)) {
        throw new Error(`Contract at ${address} doesn't look like an administered ERC 1967 proxy`);
    }
    return parseAddress(storage);
}
exports.getImplementationAddress = getImplementationAddress;
async function getBeaconAddress(provider, address) {
    const storage = await getStorageFallback(provider, address, toEip1967Hash('eip1967.proxy.beacon'));
    if (isEmptySlot(storage)) {
        throw new Error(`Contract at ${address} doesn't look like an ERC 1967 beacon proxy`);
    }
    return parseAddress(storage);
}
exports.getBeaconAddress = getBeaconAddress;
async function getStorageFallback(provider, address, ...slots) {
    let storage = '0x0000000000000000000000000000000000000000000000000000000000000000'; // default: empty slot
    for (const slot of slots) {
        storage = await (0, provider_1.getStorageAt)(provider, address, slot);
        if (!isEmptySlot(storage)) {
            break;
        }
    }
    return storage;
}
function toFallbackEip1967Hash(label) {
    return '0x' + (0, ethereumjs_util_1.keccak256)(Buffer.from(label)).toString('hex');
}
exports.toFallbackEip1967Hash = toFallbackEip1967Hash;
function toEip1967Hash(label) {
    const hash = (0, ethereumjs_util_1.keccak256)(Buffer.from(label));
    const bigNumber = new bn_js_1.default(hash).sub(new bn_js_1.default(1));
    return '0x' + bigNumber.toString(16);
}
exports.toEip1967Hash = toEip1967Hash;
function isEmptySlot(storage) {
    storage = storage.replace(/^0x/, '');
    return new bn_js_1.default(storage, 'hex').isZero();
}
function parseAddress(storage) {
    const buf = Buffer.from(storage.replace(/^0x/, ''), 'hex');
    if (!buf.slice(0, 12).equals(Buffer.alloc(12, 0))) {
        throw new Error(`Value in storage is not an address (${storage})`);
    }
    const address = '0x' + buf.toString('hex', 12, 32); // grab the last 20 bytes
    return (0, ethereumjs_util_1.toChecksumAddress)(address);
}
